src/credentials/GenerateChipX509Cert.cpp

```diff
@@ -0,0 +1,461 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements methods for generating CHIP X.509 certificate.
+ *
+ */
+
+#ifndef __STDC_LIMIT_MACROS
+#define __STDC_LIMIT_MACROS
+#endif
+
+#include <inttypes.h>
+#include <stddef.h>
+
+#include <asn1/ASN1.h>
+#include <asn1/ASN1Macros.h>
+#include <core/CHIPCore.h>
+#include <core/CHIPSafeCasts.h>
+#include <credentials/CHIPCert.h>
+#include <protocols/Protocols.h>
+#include <support/CodeUtils.h>
+#include <support/DLLUtil.h>
+
+namespace chip {
+namespace Credentials {
+
+using namespace chip::ASN1;
+using namespace chip::Protocols;
+
+struct ChipDNParams
+{
+    OID AttrOID;
+    uint64_t Value;
+};
+
+enum IsCACert
+{
+    kCACert,
+    kNotCACert,
+};
+
+static constexpr uint8_t kSHA1_Hash_Langth = 20;
```

Comment:
nit: if you place stuff into an anonymous namespace (`namespace {....}`) we can avoid all these 'static' prefixes in constants and functions.

Reply:
Good idea. I've updated the PR with the changes.